com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/ProcessConsole.java

```diff
@@ -44,76 +44,126 @@ public ProcessConsole(Process process) {
+        Observable<ConsoleMessage> stdout = this.stdoutStream.messages().map((message) -> new ConsoleMessage(message, Category.stdout));
+        Observable<ConsoleMessage> stderr = this.stderrStream.messages().map((message) -> new ConsoleMessage(message, Category.stderr));
+        this.observable = Observable.mergeArrayDelayError(stdout, stderr).observeOn(Schedulers.newThread());
     }
 
     /**
-     * Start two separate threads to monitor the messages from stdout and stderr streams of the target process.
+     * Start monitoring the process stdout/stderr streams.
      */
     public void start() {
-        this.stdoutThread = new Thread(this.name + " Stdout Handler") {
-            public void run() {
-                monitor(process.getInputStream(), stdoutSubject);
-            }
-        };
-        stdoutThread.setDaemon(true);
-        stdoutThread.start();
-
-        this.stderrThread = new Thread(this.name + " Stderr Handler") {
-            public void run() {
-                monitor(process.getErrorStream(), stderrSubject);
-            }
-        };
-        stderrThread.setDaemon(true);
-        stderrThread.start();
+        stdoutStream.start();
+        stderrStream.start();
     }
 
     /**
-     * Stop the process console handlers.
+     * Stop monitoring the process console.
      */
     public void stop() {
-        if (this.stdoutThread != null) {
-            this.stdoutThread.interrupt();
-            this.stdoutThread = null;
-        }
+        stdoutStream.stop();
+        stderrStream.stop();
+    }
 
-        if (this.stderrThread != null) {
-            this.stderrThread.interrupt();
-            this.stderrThread = null;
-        }
+    public Observable<ConsoleMessage> messages() {
+        return observable;
     }
 
-    public void onStdout(Consumer<String> callback) {
-        stdoutSubject.subscribe(callback);
+    public Observable<ConsoleMessage> stdoutMessages() {
+        return this.messages().filter((message) -> message.category == Category.stdout);
     }
 
-    public void onStderr(Consumer<String> callback) {
-        stderrSubject.subscribe(callback);
+    public Observable<ConsoleMessage> stderrMessages() {
+        return this.messages().filter((message) -> message.category == Category.stderr);
     }
 
-    private void monitor(InputStream input, PublishSubject<String> subject) {
-        BufferedReader reader = new BufferedReader(new InputStreamReader(input, encoding));
-        final int BUFFERSIZE = 4096;
-        char[] buffer = new char[BUFFERSIZE];
-        while (true) {
-            try {
-                if (Thread.interrupted()) {
-                    subject.onComplete();
-                    return;
+    /**
+     * Split the stdio message to lines, and return them as a new Observable.
+     */
+    public Observable<ConsoleMessage> lineMessages() {
+        return this.messages().map((message) -> {
+            String[] lines = message.output.split("(?<=\n)");
```

Comment:
what's the mean of (?<=\n)?

Reply:
this is a positive lookbehind regex. Able to split lines and keep delimiter.